src/unity/toolkits/object_detection/one_shot_object_detection/util/superposition.cpp

```diff
@@ -39,44 +41,43 @@ void superimpose_image(const boost::gil::rgb8_image_t::view_t &masked,
-  );
-  boost::gil::rgba8_image_t starter_image_rgba(boost::gil::rgba8_image_t::point_t(starter_image_view.dimensions()));
-  // The following line uses the color_convert conversion to convert the 
-  // background from RGB to RGBA
-  boost::gil::copy_and_convert_pixels(
-    starter_image_view,
-    view(starter_image_rgba)
-  );
-  resample_pixels(view(starter_image_rgba), transformed, M, boost::gil::bilinear_sampler());
-  color_quadrilateral(transformed, parameter_sampler.get_warped_corners());
-  superimpose_image(masked, transformed, background);
-}
-
-void superimpose_rgba_object_image(ParameterSampler &parameter_sampler,
-                                   const flex_image &object,
-                                   const boost::gil::rgb8_image_t::view_t &masked,
-                                   const boost::gil::rgba8_image_t::view_t &transformed,
-                                   const boost::gil::rgba8_image_t::view_t &background) {
+void transform_and_superimpose_object_image(ParameterSampler &parameter_sampler,
+                              const flex_image &object,
+                              const boost::gil::rgb8_image_t::view_t &superimposed,
+                              const boost::gil::rgba8_image_t::view_t &transformed,
+                              const boost::gil::rgba8_image_t::view_t &background) {
+  boost::gil::rgba8_image_t starter_image(boost::gil::rgba8_image_t::point_t(object.m_width, object.m_height));
+  if (object.m_channels == 3) {
+    // RGB
+    boost::gil::rgb8_image_t::view_t preliminary_starter_image_view = interleaved_view(
+      object.m_width,
+      object.m_height,
+      (boost::gil::rgb8_pixel_t*) (object.get_image_data()),
+      object.m_channels * object.m_width // row length in bytes
+    );
+    boost::gil::copy_and_convert_pixels(
+      preliminary_starter_image_view,
+      view(starter_image)
+    );
+  } else {
+    // RGBA
+    boost::gil::rgba8_image_t::view_t preliminary_starter_image_view = interleaved_view(
+      object.m_width,
+      object.m_height,
+      (boost::gil::rgba8_pixel_t*) (object.get_image_data()),
+      object.m_channels * object.m_width // row length in bytes
+    );
+    boost::gil::copy_pixels(
+      preliminary_starter_image_view,
+      view(starter_image)
+    );
+  }
   Eigen::Matrix<float, 3, 3> M = parameter_sampler.get_transform().inverse();
-  boost::gil::rgba8_image_t::view_t starter_image_view = interleaved_view(
-    object.m_width,
-    object.m_height,
-    (boost::gil::rgba8_pixel_t*) (object.get_image_data()),
-    object.m_channels * object.m_width // row length in bytes
-  );
-  resample_pixels(starter_image_view, transformed, M, boost::gil::bilinear_sampler());
-  superimpose_image(masked, transformed, background);
+  resample_pixels(view(starter_image), transformed, M, boost::gil::bilinear_sampler());
+  if (object.m_channels == 3) {
+    color_quadrilateral(transformed, parameter_sampler.get_warped_corners());
```

Comment:
What does this do, and why is it only needed for RGB input?

Reply:
Fixed after talking in person. Summarizing here. Turns out we need color_quadrilateral with bilinear sampling for RGBA images as well. After timing the nearest_neighbor sampling against bilinear sampling along with color_quadrilateral, the latter approach was faster and had better quality.